pageserver/src/tenant/mgr.rs

```diff
@@ -1517,21 +1521,146 @@ impl TenantManager {
+                fs::remove_dir_all(tmp_path.as_path())
+                    .await
+                    .with_context(|| format!("tenant directory {:?} deletion", tmp_path))
+            },
+        );
+
         parent_slot_guard.drop_old_value()?;
 
         // Phase 6: Release the InProgress on the parent shard
         drop(parent_slot_guard);
 
         Ok(child_shards)
     }
+
+    /// Part of [`Self::shard_split`]: hard link parent shard layers into child shards, as an optimization
+    /// to avoid the children downloading them again.
+    ///
+    /// For each resident layer in the parent shard, we will hard link it into all of the child shards.
+    async fn shard_split_hardlink(
+        &self,
+        parent_shard: &Tenant,
+        child_shards: Vec<TenantShardId>,
+    ) -> anyhow::Result<()> {
+        let parent_path = self.conf.tenant_path(parent_shard.get_tenant_shard_id());
+        let (parent_timelines, parent_layers) = {
+            let mut parent_layers = Vec::new();
+            let timelines = parent_shard.timelines.lock().unwrap().clone();
+            let parent_timelines = timelines.keys().cloned().collect::<Vec<_>>();
+            for timeline in timelines.values() {
+                let timeline_layers = timeline.get_local_layers_for_disk_usage_eviction().await;
+                for layer in timeline_layers.resident_layers.into_iter().map(|r| r.layer) {
+                    let layer = match layer {
+                        EvictionLayer::Attached(l) => l,
+                        EvictionLayer::Secondary(_) => {
+                            // Unreachable, because we fetched layers from an object of type `Tenant`, it can
+                            // only return attached layers
+                            unreachable!();
+                        }
+                    };
+
+                    let relative_path = layer
+                        .local_path()
+                        .strip_prefix(&parent_path)
+                        .context("Removing prefix from parent layer path")?;
+                    parent_layers.push(relative_path.to_owned());
+                }
+            }
+            (parent_timelines, parent_layers)
+        };
+        for child in child_shards {
+            let child_prefix = self.conf.tenant_path(&child);
+            let mut create_dirs = vec![child_prefix.clone()];
+            create_dirs.extend(
+                parent_timelines
+                    .iter()
+                    .map(|t| self.conf.timeline_path(&child, t)),
+            );
+            let layers_clone = parent_layers.clone();
+            let parent_path = parent_path.clone();
+
+            // Since we will do a large number of small filesystem metadata operations, batch them into
+            // spawn_blocking calls rather than doing each one as a tokio::fs round-trip.
+            let jh = tokio::task::spawn_blocking(move || -> anyhow::Result<()> {
+                for dir in create_dirs {
```

Comment:
```suggestion
            // spawn_blocking calls rather than doing each one as a tokio::fs round-trip.
            let span = tracing::info_span!("link_layers", child=%child.shard_id);
            let jh = tokio::task::spawn_blocking(move || -> anyhow::Result<()> {
                let _e = span.entered();
                for dir in create_dirs {
```